File: backend-beeyond/src/main/java/at/htl/beeyond/resource/CustomApplicationResource.java
package at.htl.beeyond.resource;

import at.htl.beeyond.dto.CustomApplicationDto;
import at.htl.beeyond.entity.CustomApplication;
import at.htl.beeyond.entity.User;
import at.htl.beeyond.model.FailedField;
import at.htl.beeyond.service.ValidationService;

import javax.annotation.security.RolesAllowed;
import javax.inject.Inject;
import javax.transaction.Transactional;
import javax.ws.rs.Consumes;
import javax.ws.rs.POST;
import javax.ws.rs.Path;
import javax.ws.rs.Produces;
import javax.ws.rs.core.Context;
import javax.ws.rs.core.MediaType;
import javax.ws.rs.core.Response;
import javax.ws.rs.core.SecurityContext;
import java.util.List;

@Path("/application/custom")
@Consumes(MediaType.APPLICATION_JSON)
@Produces(MediaType.APPLICATION_JSON)
public class CustomApplicationResource {

    @Inject
    ValidationService validationService;

    @POST
    @RolesAllowed({"student", "teacher"})
    @Transactional
    public Response create(@Context SecurityContext context, CustomApplicationDto customApplicationDto) {
        List<FailedField> failedFields = this.validationService.validate(customApplicationDto);
        if (!failedFields.isEmpty()) {
            return Response.status(422).entity(failedFields).build();
        }

        User user = User.find("name", context.getUserPrincipal().getName()).firstResult();
        CustomApplication customApplication = customApplicationDto.map(user);
        customApplication.persist();

        return Response.noContent().build();
    }
}

File: backend-beeyond/src/main/java/at/htl/beeyond/resource/CustomApplicationResource.kt
package at.htl.beeyond.resource

import at.htl.beeyond.dto.CustomApplicationDto
import at.htl.beeyond.entity.CustomApplication
import at.htl.beeyond.entity.User
import javax.annotation.security.RolesAllowed
import javax.transaction.Transactional
import javax.validation.Valid
import javax.ws.rs.Consumes
import javax.ws.rs.POST
import javax.ws.rs.Path
import javax.ws.rs.Produces
import javax.ws.rs.core.*

@Path("/application/custom")
@Consumes(MediaType.APPLICATION_JSON)
@Produces(MediaType.APPLICATION_JSON)
class CustomApplicationResource {

    @POST
    @RolesAllowed("student", "teacher")
    @Transactional
    fun createCustomApplication(
        @Context context: SecurityContext,
        @Valid customApplicationDto: CustomApplicationDto?,
        @Context uriInfo: UriInfo
    ): Response {
        val owner = User.find<User>("name", context.userPrincipal.name).firstResult<User>()
        val customApplication = CustomApplication(customApplicationDto, owner)

        customApplication.persist()
        val uri = uriInfo.baseUriBuilder
            .path("application")
            .path(customApplication.id.toString()).build()

        return Response.created(uri).build()
    }
}
